Source/Scene/PolygonClippingAccelerationGrid.js

```diff
@@ -0,0 +1,505 @@
+    minX = Math.min(cx, minX);
+    minY = Math.min(ay, minY);
+    minY = Math.min(by, minY);
+    minY = Math.min(cy, minY);
+
+    maxX = Math.max(ax, maxX);
+    maxX = Math.max(bx, maxX);
+    maxX = Math.max(cx, maxX);
+    maxY = Math.max(ay, maxY);
+    maxY = Math.max(by, maxY);
+    maxY = Math.max(cy, maxY);
+  }
+
+  this.topLeft = { x: minX, y: maxY };
+  this.topRight = { x: maxX, y: maxY };
+  this.btmRight = { x: maxX, y: minY };
+  this.btmLeft = { x: minX, y: minY };
+  this.width = this.topRight.x - this.topLeft.x;
+  this.height = this.topRight.y - this.btmRight.y;
+}
+
+ClippingBoundingBox2D.prototype.toClockwiseCartesian2Pairs = function () {
+  return [
+    new Cartesian2(this.topLeft.x, this.topLeft.y),
+    new Cartesian2(this.topRight.x, this.topRight.y),
+    new Cartesian2(this.btmRight.x, this.btmRight.y),
+    new Cartesian2(this.btmLeft.x, this.btmLeft.y),
+  ];
+};
+
+PolygonClippingAccelerationGrid.CellOcclusion = {
+  None: 1,
+  Partial: 2,
+  Total: 3,
+};
+
+/**
+ * Creates multiple data textures for accelerating point in polygon queries
+ * against an inputted mesh. The first data texture, 'grid', is a 2D array with (splits + 1)^2 partitions
+ * where each cell records the occlusion status of that cell with respect to all of the
+ * triangles in the inputted mesh. If a cell is completely occluded by at least
+ * one triangle in the mesh (CellOcclusion.Total) or a cell is not occluded
+ * by any triangles in the mesh (CellOcclusion.None) then the respective
+ * constant is written to grids[i * 3] where `i` is the 1D index
+ * of the 2D location in the grid array (obtained via (row * this.numCols) + col).
+ *
+ * If the cell is partially occluded by at least one triangle, then the constant
+ * CellOcclusion.Partial is written to grids[i * 3], and a list of each triangle
+ * that partially intersects this cell is inserted into 'overlappingTriangleIndices',
+ * and the inclusive / exclusive index of where to read those values is stored
+ * stored in grids[(i * 3) + 1] and grids[(i * 3) + 2] respectively.
+ *
+ * @alias PolygonClippingAccelerationGrid
+ * @constructor
+ *
+ * @param {Object} options Object with the following properties:
+ * @param {Array.<Number>} options.positions An array of position data
+ * representing the clipping mesh. Note that this function only supports
+ * generating 2D clipping data, so one dimension of data will always be
+ * thrown away.
+ * @param {Array.<Number>} options.indices Triangle index data corresponding
+ * to the positions data.
+ * @param {Number} [options.splits=32] The number of cells to place in the
+ * polygon clipping acceleration grid. Will generate (splits + 1)^2 cells.
+ * @private
+ */
+function PolygonClippingAccelerationGrid(options) {
+  this.overlappingTriangleIndices = undefined;
+  var positions = defaultValue(options.positions, []);
+  var indices = defaultValue(options.indices, []);
+  var splits = defaultValue(options.splits, 32);
```

Comment:
Doesn't matter here because it's not called frequently, but generally the better form is `var positions = defined(options.positions) ? options.positions : [];`. Avoids an unnecessary allocation.